constraints_lang.md

```diff
@@ -0,0 +1,311 @@
+
+The hope is that when a user applies `constraint`s to arguments, the arguments' unconstrained, and therefore unknown, values are filled in by the constraints.
+
+In the case of `parallel(l1, l2)`, each line argument is made up of two points, each with three components, one for each dimension. So one way to think about it is that the caller is passing in two `struct`s with 6 unknowns each. And these unknowns may be filled in by the `constraint` application.
+
+The points could also be partially constrained. I.e. the `constraint` may partially constrain values by adding relationships, but those relationships may not yet lead to fully constrained points. Additional constraints added in the future may fully constrain them.
+
+## Discussion
+
+### Using Constraints
+
+When you want to construct geometry or define your own constraint functions, you write down a series of equations. These equations take the form:
+
+```
+<expression> <relational operator> <expression>
+```
+
+For example:
+
+```
+p1.x = 5
+||l1|| = 2
+2 * pi * my_radius < travel_length + tolerance
+```
+
+Unlike definitions or assignments in imperative or functional languages, constraints allow complex expressions on either side of the operator. Since they are mathematical relationships, the left-hand-side and right-hand-side can be swapped without changing the meaning, assuming you change the relational operator appropriately.
+### Resolution Strategies
+
+Sometimes, constraints can be ambiguous. Take this example:
+
+```
+travel_length = 10
+tolerance = 0.03
+my_radious > 0
+2 * pi * my_radius < travel_length + tolerance
+```
+
+The implementation can symbolically solve for `my_radius`.
+
+```
+my_radius > 0
+my_radius < (travel_length + tolerance) / (2 * pi)
+```
+
+Because `travel_length`, `tolerance`, and `pi` are known values, we can compute an upper and lower bound for `my_radius`.
+
+However, `my_radius` is still not fully constrained. Therefore, we allow for resolution strategies.
+
+```
+travel_length = 10
+tolerance = 0.03
+my_radious > 0
+2 * pi * my_radius < travel_length
+maximize(my_radius)
+```
+
+By adding `maximize(my_radius)`, a specific value can be computed for `my_radius` even though it's not fully constrained. This is useful for displaying geometry that would only be partially constrained otherwise.
+
+Resolution strategies are only used if values are not fully constrained otherwise. You can think of them as optional constraints. This is useful for library authors while writing constraint functions. Depending on what a constraint function is applied to, the result may or may not be fully constrained. Resolution strategies allow for a way to give hints to the runtime system that aren't strict requirements. I.e. resolution strategies will never cause geometry to be over-constrained. But they can help allow unconstrained or partially constrained geometry to become fully constrained.
```

Comment:
This seems oriented (correctly) toward robustly describing 2D geometry, but I'm interested in how this maps into 3D geometry as well? This would help me understand what an "extrude" and "revolve" map to in this realm of thinking.

Reply:
I don't know. We'll have to work through more examples. Part of me thinks that it's just more geometry. I think Bézier curves are pretty straight-forward math. But this is beyond my expertise. Until we try it, I don't know what challenges we'll face. Even curves in 2D are kind of over my head. 3D is probably even more challenging.

Reply:
Yeah that's a totally fine place to be. Doing some examples as exercises seems the way to hammer those unknowns out, along with everyone else's thoughts.